### ISSUE TITLE
RFC - Extracting state machine behavior into plain old Ruby objects

### ISSUE THREAD

**Issue by bbuchalter** [solidusio/solidus#1657]
I was contemplating how we could eventually move away from the complexity and opacity of the state machines. With Solidus's commitment to sane-upgrade paths, we have to consider the smallest possible iteration, so our options are fairly limited. I was thinking that perhaps we could start by simply extracting the behavior and logic of the state machines into plain old Ruby objects which we control. 

For example:
```
# core/app/models/spree/payment/processing.rb
def capture!(amount = nil)
   ....
   started_processing!
   ....
end
```

Would become:
```
def capture!(amount = nil)
   ....
   Spree::PaymentStateMachine.new(self).started_processing!
   ....
end
```

Even this simple change has immediate benefits, as we can now clearly understand what implements `started_processing!`. Moreover, you can actually see what is done. The below implementation is accurate, but you'd never know that because state machine hides this from you.

```
module Spree
  class PaymentStateMachine
    def initialize(payment)
      @payment = payment
    end

    def started_processing!
      payment.update!(state: 'processing')
    end
  end
end
```

Maybe most importantly, we'd finally have sane error handling. For example, if the payment is in a failed state and you call `started_processing!` this is the error you get:
```
StateMachines::InvalidTransition: Cannot transition state via :started_processing from :failed (Reason(s): Payment State cannot transition via "started processing")
```

These errors are difficult to understand. And the backtraces that come with them are difficult to follow. By moving the logic the state machine implements into our own classes, we can at least _begin to consider_ how we might refactor them.

Obviously, it would be a fairly significant under taking, but also a fairly clear path of porting behaviors. If we did our jobs well, there would be no behavior changes in Solidus, simply new execution paths for the same logic.

What are the community's thoughts about the future of state machines in Solidus? I haven't seen too much conversation about this in any open issues, just https://github.com/solidusio/solidus/issues/1576 and https://github.com/solidusio/solidus/issues/266. Are there other discussions documented elsewhere?

Looking forward to thoughts about where this fits into the Solidus roadmap.

**Comment by jhawthorn** [comment]
Thanks for the thoughtful write-up Brian.

I think a lot of the complexity we'd run into re-implementing state machine as PORO would be in restricting the state transitions. We don't, for example, want to start processing failed payments. They've failed. Further actions should not be taken.

```
def started_processing!
  raise "Can't start_processing from state #{payment.state}" unless ['checkout', 'pending', 'completed'].include?(payment.state)
  payment.update!(state: 'processing')
end
```

The backtrace and error message from this might be slightly better than the one from `state_machines`, but it's still fairly awkward. Looking at the line in the backtrace will show the states which are allowed, and I suppose that could be included in the error message too. I'm not sure that's a huge gain (it doesn't help you that a "pending" payment could start processing, you have a failed payment).

I think we should also investigate continuing to use the `state_machines` gem, but defining the machine itself on a separate object.

---

The payment state machine in particular provides us almost no value in being a state machine.

![](http://i.hawth.ca/u/payment_state_full.png)

That looks vaguely like a state machine, however the transitions between `pending`, `completed`, `failed`, `void` rarely happen directly and in real usage always go through the `processing` step (search [payment/processing.rb](https://github.com/solidusio/solidus/blob/ceb1a7953ed0515b605b2b16e15989e3bb5e0428/core/app/models/spree/payment/processing.rb#L50) for `started_processing!`).

Eliminating those (usually) unused edges,

![](http://i.hawth.ca/u/payment_state_pending_only.png)

This isn't useful. Nothing prevents us from `completed` to `processing` to `pending` :bomb:. Bizarrely, the state machine doesn't stop us from transitioning from `processing` to `processing` (though at least [`handle_payment_preconditions`](https://github.com/solidusio/solidus/blob/ceb1a7953ed0515b605b2b16e15989e3bb5e0428/core/app/models/spree/payment/processing.rb#L142) does).

I believe the next step in improving the Payment states is to separate `processing` from the other states (it should exist alongside/as a modifier to the other states). Maybe that would make sense to do in a `PaymentStateMachine` object.

---

Something I'd really like to see some day is separating the `Order` state machine into its own (maybe PORO) object. When the Order state machine was first created, and wizard interfaces were in fashion, it was assumed that all orders would go through the same frontend checkout interface and all steps. This is no longer exclusively how Solidus orders are used. Sites may have one page checkouts, batch order creations, orders created as part of a return, subscriptions, etc. Implementing most of these involves fighting the state machine, either by replacing it entirely, or emulating advancing through the steps (`while order.next`, `order.advance`).

Separating the Order's state machine into a separate object would be the first step in allowing the checkout flow to be pluggable on a per-order basis.

Likely, this will be complicated, especially with the current state of the Order state machine. I think we'll need to reduce the huge number of [state machine callbacks](https://github.com/solidusio/solidus/blob/master/core/app/models/spree/order/checkout.rb) before we can attempt this.

---

TL;DR. I like it, but complicated.

**Comment by bbuchalter** [comment]
@jhawthorn thanks for your thoughtful reply. It's good for us to get this conversation started. Some specific replies below:

> I think a lot of the complexity we'd run into re-implementing state machine as PORO would be in restricting the state transitions.

I had considered this, but honestly it felt like one of the _easiest_ parts of the extraction. You captured it perfectly in your code snippet. Is there more to it?

> Separating the Order's state machine into a separate object would be the first step in allowing the checkout flow to be pluggable on a per-order basis.

Do you see separating the state machines from the objects they mutate implemented as an extension point, similar to what's been done recently for tax calculations?

> Likely, this will be complicated, especially with the current state of the Order state machine. I think we'll need to reduce the huge number of state machine callbacks before we can attempt this.

Wow! I hadn't seen that Checkout state machine in a long time, and it hasn't gotten any better.  However, looking more closely, it's really just a bunch of method calls executed at a particular transition, in a particular order. I think extracting it out into a class which calls those methods in the correct order (and under direct test!) would be very possible. What do you think?

**Comment by alepore** [comment]
maybe relevant, a state machine implemented with classes: https://github.com/gocardless/statesman

**Comment by tvdeyen** [comment]
> Statesman is extremely opinionated and is designed to provide a robust audit trail and data integrity. It decouples the state machine logic from the underlying model and allows for easy composition with one or more model classes.

@alepore this looks very promising. Thanks for sharing.

**Comment by dhonig** [comment]
Is there any advantage to StatesMan other than what appears to be a much better API, it is extremely similar to what is already there via the state_machine gem, or am I missing something?

**Comment by BenMorganIO** [comment]
Just my 2 cents. A friend and I once took the time to try to build a payments gem called Ryal. Largely an experiment. We wrote the state machine for it three times and constantly consulted an accountant for iterating over how a payment process should ideally work.

This was the first try with AASM: https://github.com/spurryal/spur/pull/5
Second attempt with Statesman: https://github.com/spurryal/spur/pull/9
Third and final attempt with AASM and ideas pulled from Statesman: https://github.com/spurryal/spur/pull/10

Basic AASM was used in the beginning, but we tried out statesman. We ended up feeling that it had a lot of "bonus" features, wasn't as "essentially" featureful as AASM, so we stuck to AASM. All the bonus features inside of Statesmen we implemented ourselves, such as the caching of the transitions.

You can view the readme and what we ended up finalizing on as a product here: https://github.com/spurryal/spur/tree/master/ryal#payment-state-machine

What was really cool was that we added what we called a "machinist". This was a module where we placed state machine specific classes and then called back to them via a `#machinist` method. You can see the state machine we build [here](https://github.com/spurryal/spur/blob/master/ryal/app/models/spur/machinist/payment.rb) and how we accessed it [here](https://github.com/spurryal/spur/blob/master/ryal/app/models/spur/payment.rb#L25).

All in all:

When it came to the state machine, we did end up building what we wanted, but if we could refactor it, we'd keep our current tests (as we quite liked the end result behavior wise) and ever so slowly start removing AASM and write our own ruby methods. Our main goal was to have a trustworthy state machine however to get that, it required a lot of heading back to the drawing board.

What I'd do with state machines with Solidus:

- I would pull the current state machines all into their own "machinest" models.
- Ensure they all behave similarly. Check all their behavior from their usage of `!` methods down to if they cache the transitions. If they don't, then this will most likely be the second stage of behavioral changes.
- Since all the behavior has been outlined and they're all in one place, I'd leave AASM as a dependency and move some of the methods out into their own ruby code, leaving AASM with less code it owns.
- If we can, consult industry professionals and revise what some state machines should look like. Although this will lead to some major business logic changes, it would most likely be welcomed by some power users and help align with prospective business users. We loved it that we could go back and forth with an accountant. The state machine we build looks fairly straightforward, but that's only because we spent the time to make it simple and efficient. Simplicity was hard work.
- The big removal AASM and providing an upgrade path for those who are dependent on extending the current Solidus state machine with AASM.

Hope this helps in considering the refactor of some state machines :)

**Comment by bbuchalter** [comment]
There doesn't seem to be any further comment here.
